### PR TITLE
Codegen: Query span as late as possible

### PR DESCRIPTION
**Pull request by bugadani**

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/consts.rs [hunk 1]
```diff
@@ -210,21 +212,21 @@ impl CodegenCx<'ll, 'tcx> {
...
+            let id = self.tcx.hir().local_def_id_to_hir_id(local_def_id);
             let llty = self.layout_of(ty).llvm_type(self);
             // FIXME: refactor this to work without accessing the HIR
             let (g, attrs) = match self.tcx.hir().get(id) {
-                Node::Item(&hir::Item { attrs, span, kind: hir::ItemKind::Static(..), .. }) => {
+                Node::Item(&hir::Item { attrs, kind: hir::ItemKind::Static(..), .. }) => {
                     if let Some(g) = self.get_declared_value(sym) {
                         if self.val_ty(g) != self.type_ptr_to(llty) {
-                            span_bug!(span, "Conflicting types for static");
+                            span_bug!(self.tcx.def_span(def_id), "Conflicting types for static");
```
Comment (lcnr) [inline compiler/rustc_codegen_llvm/src/consts.rs:223]: is there a reason you aren't using the `Item` span here? That seems easier to me

Comment (bugadani) [reply compiler/rustc_codegen_llvm/src/consts.rs:223]: At the moment, it is, yes. Consider this change a step towards #43088 :)

Comment (bugadani) [reply compiler/rustc_codegen_llvm/src/consts.rs:223]: Also I actually dont know, by the way, if these are identical, so it's kind of a "let's see what the tests think" kind of change.

Comment (lcnr) [reply compiler/rustc_codegen_llvm/src/consts.rs:223]: the tests won't think anything here will they? This is a `span_bug` so we should new reach this in tests

Comment (bugadani) [reply compiler/rustc_codegen_llvm/src/consts.rs:223]: GitHub needs a reaction that indicates how blind I can be :)

**Comment by bugadani** [comment]
Hmm, can I?
@bors try

**Comment by lcnr** [comment]
completely forgot about this PR, sry.

Afaict perf was slightly positive and it seems good to land, so

@bors r+

**Comment by bugadani** [comment]
I'm on mobile right now but looks the same as https://github.com/rust-lang/rust/pull/78489#issuecomment-719597114

What's up with macOS?

**Comment by m-ou-se** [comment]
@rust-lang/infra ^

**Comment by kennytm** [comment]
🤔 

https://github.com/rust-lang-ci/rust/runs/1336669580

```
---- [debuginfo-lldb] debuginfo/pretty-std-collections.rs stdout ----
NOTE: compiletest thinks it is using LLDB version 1200
NOTE: compiletest thinks it is using LLDB without native rust support

error: Error while running LLDB
status: exit code: 1
command: "/usr/bin/python3" "/Users/runner/work/rust/rust/src/etc/lldb_batchmode.py" "/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/debuginfo/pretty-std-collections.lldb/a" "/Users/runner/work/rust/rust/build/x86_64-apple-darwin/test/debuginfo/pretty-std-collections.lldb/pretty-std-collections.debugger.script"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
error: need to add support for DW_TAG_base_type '()' encoded with DW_ATE = 0x7, bit_size = 0
clangclang: CommandLine Error: Option 'h' registered more than once!
: CommandLine Error: Option 'LLVM ERROR: inconsistency in registered CommandLine options
h' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options

------------------------------------------



failures:
    [debuginfo-lldb] debuginfo/pretty-std-collections.rs

test result: FAILED. 89 passed; 1 failed; 26 ignored; 0 measured; 0 filtered out
```

**Comment by bugadani** [comment]
@pietroalbini could you please retry this again? The previous one looks like it failed with a network error:

```
warning: spurious network error (2 tries remaining): failed to get 200 response from `https://crates.io/api/v1/crates/matchers/0.0.1/download`, got 502
warning: spurious network error (1 tries remaining): failed to get 200 response from `https://crates.io/api/v1/crates/matchers/0.0.1/download`, got 502
error: failed to download from `https://crates.io/api/v1/crates/matchers/0.0.1/download`
```

**Comment by bugadani** [comment]
I'm starting to think this PR is cursed :(

**Comment by pietroalbini** [comment]
~~Congrats, you found a new kind of spurious failure!~~

@bors retry

**Pull request merged**